src/js/live-tracker.js

```diff
@@ -0,0 +1,204 @@
+      }
+
+      this.pastSeekEnd_ = 0;
+      this.lastSeekEnd_ = newSeekEnd;
+      this.trigger('seekableendchange');
+    }
+
+    this.pastSeekEnd_ = this.pastSeekEnd() + 0.03;
+
+    if (this.isBehind_() !== this.behindLiveEdge()) {
+      this.behindLiveEdge_ = this.isBehind_();
+      this.trigger('liveedgechange');
+    }
+  }
+
+  /**
+   * handle a durationchange event on the player
+   * and start/stop tracking accordingly.
+   */
+  handleDurationchange() {
+    if (this.player_.duration() === Infinity) {
+      this.start();
+    } else {
+      this.stop();
+    }
+  }
+
+  /**
+   * start tracking live playback
+   */
+  start() {
+    if (this.started()) {
+      return;
+    }
+
+    this.trackingInterval_ = this.setInterval(this.trackLive_, 30);
+    this.trackLive_();
+  }
+
+  /**
+   * stop tracking live playback
+   */
+  stop() {
+    if (!this.started()) {
+      return;
+    }
+    this.pastSeekEnd_ = 0;
+    this.lastSeekEnd_ = null;
+    this.behindLiveEdge_ = null;
+
+    this.clearInterval(this.trackingInterval_);
+    this.trackingInterval_ = null;
+  }
+
+  /**
+   * A helper to get the player seekable end
+   * so that we don't have to null check everywhere
+   */
+  seekableEnd() {
+    const seekable = this.player_.seekable();
+
+    if (!seekable || !seekable.length) {
+      return Infinity;
+    }
+
+    return seekable.end(0);
+  }
+
+  /**
+   * A helper to get the player seekable start
+   * so that we don't have to null check everywhere
+   */
+  seekableStart() {
+    const seekable = this.player_.seekable();
+
+    if (!seekable || !seekable.length) {
+      return 0;
+    }
+
+    return seekable.start(0);
+  }
+
+  /**
+   * Get the live time window
+   */
+  liveWindow() {
+    return this.seekableEnd() - this.seekableStart();
+  }
+
+  /**
+   * Determines if the player is live, only checks if this component
+   * is tracking live playback or not
+   */
+  isLive() {
+    return this.started();
+  }
+
+  /**
+   * Determines if currentTime is at the live edge and won't fall behind
+   * on each seekableendchange
+   */
+  atLiveEdge() {
+    return !this.behindLiveEdge();
+  }
+
+  /**
+   * get what we expect the live current time to be
+   */
+  liveCurrentTime() {
+    return this.pastSeekEnd() + this.seekableEnd();
+  }
+
+  /**
+   * Returns how far past seek end we expect current time to be
+   */
+  pastSeekEnd() {
+    return this.pastSeekEnd_ || 0;
+  }
+
+  /**
+   * If we are currently behind the live edge, aka currentTime will be
+   * behind on a seekableendchange
+   */
+  behindLiveEdge() {
+    return this.behindLiveEdge_;
+  }
+
+  started() {
+    return typeof this.trackingInterval_ === 'number';
+  }
+
+  /**
+   * Seek to the live edge if we are behind the live edge
+   */
+  seekToLiveEdge() {
+    if (this.atLiveEdge()) {
+      return;
+    }
+
+    this.player().pause();
+    this.player().addClass('vjs-waiting');
+    this.one('seekableendchange', () => {
```

Comment:
why wait for `seekableendchange` for `seekToLiveEdge`? Shouldn't just set `currentTime(this.seekableEnd())`?

Reply:
So we have to wait for `seekableendchange` because you can only seek within `seekable`. Example:

* You are at 4s 
* seekable end is currently at 8s
* seekable has 4s added to it every 4s (I called this seekableIncrement_)
* live playback is a bit passed seekable end, say 10s


If we seek right away you will appear to be live until seekable end changes, then you will be 2s behind, and catch up after 2s. Then you will be 2s behind again when seekable end changes again.

Reply:
I'm not sure I follow. Also, what happens if we had just missed the previous seekable end change? Would we have to wait 4 seconds to get to the live edge?

Reply:
yes it would

Reply:
As a user I would be very confused and think that something is broken. @videojs/playback any ideas on how to better handle this?

Reply:
I see it change from at live edge and not at live edge as part of regular playback.

Reply:
We may want to be careful with how we define live. If segments are 4 seconds long, then we probably want live to be considered within the last 8 seconds (2 segment durations), because of the case mentioned (where, if we are unlucky with manifest refreshes, we could end up seeking to a segment we think is at the live edge, but the playlist refreshes just after the request, so we start 4 seconds + request and processing duration back from what we think is seekable end).

The other solution is to seek to a point past where the seekable end is, figuring that we'll wait until the next refresh to ensure we start at the most live point, but this delays startup up to the duration of a segment (plus request time). I think this experience would be worse, but can be controlled through shorter segment durations.

Reply:
Could we just keep track of timestamps for when we get seekableendchange events, and then seek to `this.seekableEnd() + timeSinceLastSeekableendchange`?